src/wav.rs

```diff
@@ -0,0 +1,388 @@
+        let tag = tag.try_into()?;
+        Ok(Self(tag))
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+struct ChunkHeader {
+    tag: ChunkTag,
+    size: u32,
+}
+
+impl ChunkHeader {
+    /// Reads a RIFF header from the input stream. Such header is composed of:
+    ///
+    /// | Field    | Size | Value                         |
+    /// |----------+------+-------------------------------|
+    /// | RIFF tag |    4 | "RIFF"                        |
+    /// | size     |    4 | 32 bits little endian integer |
+    /// | WAVE tag |    4 | "WAVE"                        |
+    pub fn read_riff_header<R>(mut reader: R) -> crate::Result<ChunkHeader>
+    where
+        R: io::Read,
+    {
+        const RIFF: ChunkTag = ChunkTag(*b"RIFF");
+        const WAVE: ChunkTag = ChunkTag(*b"WAVE");
+
+        let invalid_header_error = Error::new(ErrorKind::InvalidInput, "invalid WAV/RIFF header");
+
+        const BUFFER_SIZE: usize = (CHUNK_HEADER_LEN + WAVE_TAG.len()) as usize;
+
+        let mut buffer = [0; BUFFER_SIZE];
+
+        // Use a single read call to improve performance on unbuffered readers.
+        reader.read_exact(&mut buffer)?;
+
+        let chunk_header: ChunkHeader = buffer[0..8]
+            .try_into()
+            .expect("slice with incorrect length");
+
+        if !chunk_header.tag.is(&RIFF) {
+            return Err(invalid_header_error);
+        }
+
+        let wave_tag: ChunkTag = buffer[8..12]
+            .try_into()
+            .expect("slice with incorrect length");
+
+        if !wave_tag.is(&WAVE) {
+            return Err(invalid_header_error);
+        }
+
+        Ok(chunk_header)
+    }
+
+    /// Reads a chunk header from the input stream. A header is composed of:
+    ///
+    /// | Field | Size | Value                         |
+    /// |-------+------+-------------------------------|
+    /// | tag   |    4 | chunk type                    |
+    /// | size  |    4 | 32 bits little endian integer |
+    pub fn read<R>(mut reader: R) -> io::Result<Self>
+    where
+        R: io::Read,
+    {
+        const BUFFER_SIZE: usize = CHUNK_HEADER_LEN as usize;
+
+        let mut header = [0; BUFFER_SIZE];
+
+        // Use a single read call to improve performance on unbuffered readers.
+        reader.read_exact(&mut header)?;
+
+        Ok(header.into())
+    }
+
+    /// Finds an ID3 chunk in a flat sequence of chunks. This should be called after reading
+    /// the root RIFF chunk.
+    ///
+    /// # Arguments
+    ///
+    /// * `reader` - The input stream. The reader must be positioned right after the RIFF
+    ///              chunk header.
+    /// * `end` - The stream position where the chunk sequence ends. This is used to
+    ///           prevent searching past the end.
+    pub fn find_id3<R>(reader: R, end: u64) -> crate::Result<Self>
+    where
+        R: io::Read + io::Seek,
+    {
+        const ID3: ChunkTag = ChunkTag(*b"ID3 ");
```

Comment:
We can remove this const in favour of the module scoped tag

Reply:
Oops, missed that one. I'll fix it right away.